cmd/convertor/builder/builder.go

```diff
@@ -95,41 +97,167 @@ func NewOverlayBDBuilder(ctx context.Context, opt BuilderOptions) (Builder, erro
 			Transport: transport,
 		},
 	})
-	engineBase, err := getBuilderEngineBase(ctx, resolver, opt.Ref, opt.TargetRef)
+	fetcher, err := resolver.Fetcher(ctx, opt.Ref)
 	if err != nil {
-		return nil, err
+		return fmt.Errorf("failed to new fetcher: %w", err)
 	}
-	engineBase.workDir = opt.WorkDir
-	engineBase.oci = opt.OCI
-	engineBase.mkfs = opt.Mkfs
-	engineBase.vsize = opt.Vsize
-	engineBase.db = opt.DB
 
-	refspec, err := reference.Parse(opt.Ref)
+	_, src, err := resolver.Resolve(ctx, opt.Ref)
 	if err != nil {
-		return nil, err
+		return fmt.Errorf("failed to resolve: %w", err)
+	}
+	var manifestDescs []v1.Descriptor
+	switch src.MediaType {
+	case v1.MediaTypeImageManifest, images.MediaTypeDockerSchema2Manifest:
+		manifestDescs = append(manifestDescs, src)
+	case v1.MediaTypeImageIndex, images.MediaTypeDockerSchema2ManifestList:
+		rc, err := fetcher.Fetch(ctx, src)
+		if err != nil {
+			return fmt.Errorf("failed to fetch index: %w", err)
+		}
+		defer rc.Close()
+		b, err := io.ReadAll(rc)
+		if err != nil {
+			return fmt.Errorf("failed to read index: %w", err)
+		}
+		var index v1.Index
+		if err := json.Unmarshal(b, &index); err != nil {
+			return fmt.Errorf("failed to parse index: %w", err)
+		}
+		// (TODO) platform filter
+		manifestDescs = append(manifestDescs, index.Manifests...)
```

Comment:
it may be possible that an index has another index which has other manifests? maybe good to handle ImageIndex case recursively?

Reply:
This is in fact explicitly valid according to the oci spec https://github.com/opencontainers/image-spec/blob/f5f87016de46439ccf91b5381cf76faaae2bc28f/image-index.md?plain=1#L46 I doubt its common common though or if wed want to support it in practice.

Reply:
Anyway, I rewrote a version of the implementation to handle the recursive index.